packages/oauth/oauth-client/src/oauth-client-auth.ts

```diff
@@ -0,0 +1,152 @@
+    // more robust against keysets containing keys without a "kid" property.
+    for (const key of keyset.list({ use: 'sig', alg })) {
+      // Return the first key from the key set that matches the server's
+      // supported algorithms.
+      if (key.isPrivate && key.kid) {
+        return { method: 'private_key_jwt', kid: key.kid }
+      }
+    }
+
+    throw new Error(
+      alg.includes(FALLBACK_ALG)
+        ? `Client authentication method "${method}" requires at least one "${FALLBACK_ALG}" signing key with a "kid" property`
+        : // AS is not compliant with the ATproto OAuth spec.
+          `Authorization server requires "${method}" authentication method, but does not support "${FALLBACK_ALG}" algorithm.`,
+    )
+  }
+
+  if (method === 'none') {
+    return { method: 'none' }
+  }
+
+  throw new Error(
+    `The ATProto OAuth spec requires that client use either "none" or "private_key_jwt" authentication method.` +
+      (method === 'client_secret_basic'
+        ? ' You might want to explicitly set "token_endpoint_auth_method" to one of those values in the client metadata document.'
+        : ` You set "${method}" which is not allowed.`),
+  )
+}
+
+export type ClientCredentialsFactory = () => Awaitable<OAuthClientCredentials>
+
+/**
+ * @throws {AuthMethodUnsatisfiableError} if the authentication method is no
+ * long usable (either because the AS changed, of because the key is no longer
+ * available in the keyset).
+ */
+export function createClientCredentialsFactory(
+  authMethod: ClientAuthMethod,
+  serverMetadata: OAuthAuthorizationServerMetadata,
+  clientMetadata: ClientMetadata,
+  runtime: Runtime,
+  keyset?: Keyset,
+): ClientCredentialsFactory {
+  // Ensure the AS still supports the auth method.
+  if (!supportedMethods(serverMetadata).includes(authMethod.method)) {
+    throw new AuthMethodUnsatisfiableError(
+      `Client authentication method "${authMethod.method}" no longer supported`,
+    )
+  }
+
+  if (authMethod.method === 'none') {
+    return () => ({
+      client_id: clientMetadata.client_id,
+    })
+  }
+
+  if (authMethod.method === 'private_key_jwt') {
+    try {
+      // The client used to be a confidential client but no longer has a keyset.
+      if (!keyset) throw new Error('A keyset is required for private_key_jwt')
+
+      // @NOTE throws if no matching key can be found
+      const [key, alg] = keyset.findPrivateKey({
+        use: 'sig',
+        kid: authMethod.kid,
+        alg: supportedAlgs(serverMetadata),
+      })
+
+      return async () => ({
+        client_id: clientMetadata.client_id,
+        client_assertion_type: CLIENT_ASSERTION_TYPE_JWT_BEARER,
+        client_assertion: await key.createJwt(
+          { alg },
+          {
+            iss: clientMetadata.client_id,
+            sub: clientMetadata.client_id,
+            aud: serverMetadata.issuer,
+            jti: await runtime.generateNonce(),
+            iat: Math.floor(Date.now() / 1000),
+          },
```

Comment:
I think per the spec this must have an `exp`.

Reply:
Are you sure? I don't see it here:
https://datatracker.ietf.org/doc/html/rfc9449#section-4.2